lib/fastlane/plugin/firebase_app_distribution/client/firebase_app_distribution_api_client.rb

```diff
@@ -0,0 +1,148 @@
+      def upload(app_id, binary_path)
+        upload_token = get_upload_token(app_id, binary_path)
+        upload_status_response = get_upload_status(app_id, upload_token)
+        if upload_status_response.success?
+          UI.success("This APK/IPA has been uploaded before. Skipping upload step.")
+        else
+          UI.message("This APK has not been uploaded before.")
+          MAX_POLLING_RETRIES.times do
+            if upload_status_response.success?
+              UI.success("Uploaded APK/IPA Successfully!")
+              break
+            elsif upload_status_response.in_progress?
+              sleep(POLLING_INTERVAL_SECONDS)
+            else
+              UI.message("Uploading the APK/IPA.")
+              upload_binary(app_id, binary_path)
+            end
+            upload_status_response = get_upload_status(app_id, upload_token)
+          end
+          unless upload_status_response.success?
+            UI.error("It took longer than expected to process your APK/IPA, please try again.")
+            return nil
+          end
+        end
+        upload_status_response.release_id
+      end
+
+      # Gets the upload status for the app release.
+      def get_upload_status(app_id, app_token)
+        begin
+          response = connection.get(upload_status_url(app_id, app_token)) do |request|
+            request.headers["Authorization"] = "Bearer " + auth_token
+          end
+        rescue Faraday::ResourceNotFound
+          UI.crash!("#{ErrorMessage::INVALID_APP_ID}: #{app_id}")
+        end
+        return UploadStatusResponse.new(response.body)
+      end
+
+      private
+
+      def v1_apps_url(app_id)
+        "/v1alpha/apps/#{app_id}"
+      end
+
+      def release_notes_create_url(app_id, release_id)
+        "#{v1_apps_url(app_id)}/releases/#{release_id}/notes"
+      end
+
+      def binary_upload_url(app_id)
+        "/app-binary-uploads?app_id=#{app_id}"
+      end
+
+      def upload_status_url(app_id, app_token)
+        "#{v1_apps_url(app_id)}/upload_status/#{app_token}"
+      end
+
+      def upload_token_format(app_id, project_number, binary_hash)
+        CGI.escape("projects/#{project_number}/apps/#{app_id}/releases/-/binaries/#{binary_hash}")
+      end
+
+      def connection
+        @connection ||= Faraday.new(url: Helper::FirebaseAppDistributionHelper::BASE_URL) do |conn|
+          conn.response(:json, parser_options: { symbolize_names: true })
+          conn.response(:raise_error) # raise_error middleware will run before the json middleware
+          conn.adapter(Faraday.default_adapter)
+        end
+      end
+
+      def auth_token
+        @auth_token ||= begin
+          client = Signet::OAuth2::Client.new(
+            token_credential_uri: TOKEN_CREDENTIAL_URI,
+            client_id: Fastlane::Actions::FirebaseAppDistributionLoginAction::CLIENT_ID,
+            client_secret: Fastlane::Actions::FirebaseAppDistributionLoginAction::CLIENT_SECRET,
```

Comment:
Do `CLIENT_SECRET` and `CLIENT_ID` still live in the Action? Should they be here in the client class?

Reply:
They're in the LoginAction and also used there for authentication

https://github.com/fastlane/fastlane-plugin-firebase_app_distribution/blob/baeb25cd03c243a9a7446b5b804bf44ca8e7ffd8/lib/fastlane/plugin/firebase_app_distribution/actions/firebase_app_distribution_login.rb#L11-L17

Since our auth/login is going to change, where do you think should be the best place for this?

Reply:
Ahh that makes sense. I wonder if logging in should be handled by the API? I don't think it's necessary to make that decision right now but definitely something to think about as we add more ways to authenticate.

Reply:
Ah, I see. Agreed, feel free to leave for now if you want.